static/js/tools/map/time_slider.tsx

```diff
@@ -0,0 +1,227 @@
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Time slider component.
+ */
+
+import React, { useEffect, useRef, useState } from "react";
+
+interface TimeSliderProps {
+  // Current date of map
+  currentDate: string;
+
+  // Selected dates to display on time slider
+  dates: Array<string>;
+
+  // Hash representing current time series
+  metahash: string;
+
+  // Whether the slider is enabled on refresh
+  // False if the map is displaying multiple dates (e.g. 'Best Available')
+  startEnabled: boolean;
+
+  // Fetches data for slider dates when play is pressed
+  onPlay(metahash: string, callback: () => void): any;
+
+  // Updates map date to slider date
+  updateDate(metahash: string, date: string): any;
+}
+
+export function TimeSlider(props: TimeSliderProps): JSX.Element {
+  const INTERVAL_MS = 500;
+  const SLIDER_MARGIN = 16;
+  const HANDLE_WIDTH = 4;
+  const HANDLE_MARGIN = SLIDER_MARGIN - HANDLE_WIDTH / 2;
+
+  const start = props.dates[0];
+  const end = props.dates[props.dates.length - 1];
+
+  const [currentDate, setCurrentDate] = useState(
+    props.startEnabled ? props.currentDate : "--"
+  );
+  const [enabled, setEnabled] = useState(props.startEnabled);
+  const [index, setIndex] = useState(
+    props.startEnabled ? getIndex(props.dates, props.currentDate) : -1
+  );
+  const [loaded, setLoaded] = useState(false);
+
+  // Number of pixels the handle is offset from the left edge of the slider bar
+  const [offset, setOffset] = useState(null);
+  const [play, setPlay] = useState(true);
+  const [timer, setTimer] = useState(null);
+
+  const firstUpdate = useRef(true);
+
+  useEffect(() => {
+    setLoaded(false);
+    setIndex(
+      props.startEnabled ? getIndex(props.dates, props.currentDate) : -1
+    );
+    setCurrentDate(props.startEnabled ? props.currentDate : "--");
+  }, [props.metahash]);
+
+  useEffect(() => {
+    if (enabled) {
+      setOffset(
+        getOffset(start, end, currentDate, SLIDER_MARGIN, HANDLE_MARGIN)
+      );
+    }
+  }, [currentDate, getOffset]);
+
+  useEffect(() => {
+    function handleResize() {
+      if (enabled) {
+        setOffset(
+          getOffset(start, end, currentDate, SLIDER_MARGIN, HANDLE_MARGIN)
+        );
+      }
+    }
+    window.addEventListener("resize", handleResize);
+  });
+
+  useEffect(() => {
+    // Don't update pre-selected date until user presses play
+    if (firstUpdate.current) {
+      firstUpdate.current = false;
+      return;
+    }
+    if (index >= 0) {
+      setCurrentDate(props.dates[index]);
+      if (loaded) {
+        props.updateDate(props.metahash, props.dates[index]);
+      }
+    }
+    setEnabled(true);
+    if (index == props.dates.length - 1) {
```

Comment:
in typescript, for equals, we want to use "===" because "==" doesn't check datatypes. Same everywhere else in this PR.

Reply:
done